openlibrary/plugins/openlibrary/js/ile/utils/SelectionManager/SelectionManager.js

```diff
@@ -82,19 +84,77 @@ export default class SelectionManager {
     /**
      * @param {MouseEvent & { currentTarget: HTMLElement }} clickEvent
      */
-    toggleSelected(clickEvent) {
+    processClick(clickEvent) {
         // If there is text selection or the click is on a link that isn't a select handle, don't do anything
-        if (window.getSelection()?.toString() !== '' ||
+        if ((!clickEvent.shiftKey && window.getSelection()?.toString() !== '') ||
             ($(clickEvent.target).closest('a').is('a') &&
             $(clickEvent.target).not('.ile-select-handle').length > 0)) return;
 
         const el = clickEvent.currentTarget;
+        if (clickEvent.shiftKey && this.lastClicked)
+        {
+            this.clearTextSelection();
+            const siblingSet = this.getSiblingSet(el);
+            const lastClickedIndex = siblingSet.index(this.lastClicked);
+            const elIndex = siblingSet.index(el);
+            if (lastClickedIndex > -1 && Math.abs(elIndex - lastClickedIndex) > 1) {
+                let affectedElements;
+                if (elIndex > lastClickedIndex) {
+                    affectedElements = siblingSet.slice(lastClickedIndex + 1, elIndex + 1);
+                } else {
+                    affectedElements = siblingSet.slice(elIndex, lastClickedIndex);
+                }
+                const stateChange = el.classList.contains('ile-selected') ? false : true;
+                for (const element of affectedElements) this.toggleSelected(element, stateChange);
+            } else {
+                this.toggleSelected(el);
+            }
+        }
+        else {
+            this.toggleSelected(el);
+        }
+        this.lastClicked = el;
+        this.updateToolbar();
+    }
+
+    /**
+     * Sets of selectable elements are sometimes HTML siblings and sometimes not. This function
+     * hides that complexity by finding the common parent between the passed element and the last
+     * clicked element and generating a set of siblings from that information
+     *
+     * @param {HTMLElement} clicked
+     */
+    getSiblingSet(clicked) {
+        const lastParents = $(this.lastClicked).parents();
+        const clickedParents = $(clicked).parents();
+        let commonParent = undefined;
+        for (let i = 0; i < 2 && i < lastParents.length; i++) {
```

Comment:
Can do this a little cleaner as:

```js
let commonParent = undefined;
let curEls = { clicked, lastClicked: this.lastClicked };
// Only check up to 2 levels up in the tree
for (let i = 0; i < 2; i++) {
    if (!curEls.clicked || !curEls.lastClicked) {
        break;
    } else if (curEls.clicked == curEls.lastClicked) {
        commonParent = curEls.clicked;
    } else {
        curEls.clicked = curEls.clicked.parentElement;
        curEls.lastClicked = curEls.lastClicked.parentElement;
    }
}
```

This avoids some of the iterator maths and I think more conceptually matches the approach (walk up the tree from the two nodes until a common parent is found).